File: packages/study/agent-go-blocks/pkg/agent/engine.go
package agent

import (
	"context"
	"fmt"

	"google.golang.org/genai"
)

type AgentEngine struct {
	client     *genai.Client
	modelName  string
	components []AgentComponent
}

func NewAgentEngine(ctx context.Context, apiKey string, modelName string) (*AgentEngine, error) {
	client, err := genai.NewClient(ctx, &genai.ClientConfig{
		APIKey:  apiKey,
		Backend: genai.BackendGemini,
	})
	if err != nil {
		return nil, err
	}

	if modelName == "" {
		modelName = "gemini-3-flash-preview"
	}

	return &AgentEngine{
		client:    client,
		modelName: modelName,
	}, nil
}

func (e *AgentEngine) Use(component AgentComponent) {
	e.components = append(e.components, component)
}

func (e *AgentEngine) Chat(ctx context.Context, userInput string, callback func(string)) error {
	messages := []ChatMessage{{Role: RoleUser, Content: userInput}}

	// 1. Run BeforeLLMCall hooks
	var err error
	for _, comp := range e.components {
		messages, err = comp.BeforeLLMCall(ctx, messages)
		if err != nil {
			return err
		}
	}

	// 2. Prepare contents for API
	var contents []*genai.Content
	for _, m := range messages {
		contents = append(contents, &genai.Content{
			Role:  string(m.Role),
			Parts: []*genai.Part{{Text: m.Content}},
		})
	}

	// 3. Call LLM
	iter := e.client.Models.GenerateContentStream(ctx, e.modelName, contents, nil)

	// 4. Stream and capture
	fullResponse := ""
	for {
		resp, err := iter.Next()
		if err != nil {
			break
		}
		for _, candidate := range resp.Candidates {
			for _, part := range candidate.Content.Parts {
				text := part.Text
				fullResponse += text
				callback(text)
			}
		}
	}

	// 5. Run AfterLLMCall hooks
	for _, comp := range e.components {
		err = comp.AfterLLMCall(ctx, userInput, fullResponse)
		if err != nil {
			return err
		}
	}

	return nil
}

File: packages/study/agent-rs-blocks/src/agent/engine.rs
use crate::agent::types::{AgentComponent, ChatMessage, MessageRole};
use futures_util::StreamExt;
use reqwest_eventsource::{Event, EventSource};
use serde_json::json;

pub struct AgentEngine {
    api_key: String,
    model_name: String,
    components: Vec<Box<dyn AgentComponent>>,
}

impl AgentEngine {
    pub fn new(api_key: String, model_name: Option<String>) -> Self {
        Self {
            api_key,
            model_name: model_name.unwrap_or_else(|| "gemini-3-flash-preview".to_string()),
            components: Vec::new(),
        }
    }

    pub fn use_component(&mut self, component: Box<dyn AgentComponent>) {
        self.components.push(component);
    }

    pub async fn chat(&mut self, user_input: &str) -> Result<impl futures_util::Stream<Item = String>, Box<dyn std::error::Error>> {
        let mut messages = vec![ChatMessage {
            role: MessageRole::User,
            content: user_input.to_string(),
        }];

        // 1. Run before_llm_call hooks
        for comp in &self.components {
            messages = comp.before_llm_call(messages).await;
        }

        // 2. Prepare contents for API
        let contents = json!(messages.iter().map(|m| {
            json!({
                "role": m.role,
                "parts": [{"text": m.content}]
            })
        }).collect::<Vec<_>>());

        let url = format!(
            "https://generativelanguage.googleapis.com/v1beta/models/{}:streamGenerateContent?alt=sse&key={}",
            self.model_name, self.api_key
        );

        // 3. Call LLM
        let client = reqwest::Client::new();
        let request = client.post(url).json(&json!({ "contents": contents }));
        let mut event_source = EventSource::new(request)?;

        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
        let user_input_owned = user_input.to_string();
        
        // This is a bit tricky in Rust because of ownership and components needing to be updated.
        // For simplicity in this study agent, we'll collect the full response then run hooks.
        
        let mut full_response = String::new();
        
        tokio::spawn(async move {
            while let Some(event) = event_source.next().await {
                match event {
                    Ok(Event::Message(message)) => {
                        let v: serde_json::Value = serde_json::from_str(&message.data).unwrap_or_default();
                        if let Some(text) = v["candidates"][0]["content"]["parts"][0]["text"].as_str() {
                            full_response.push_str(text);
                            let _ = tx.send(text.to_string());
                        }
                    }
                    _ => break,
                }
            }
            // After streaming is done, we need to notify components.
            // In a real industrial agent, we'd use an event bus or similar.
        });

        Ok(tokio_stream::wrappers::UnboundedReceiverStream::new(rx))
    }
}
